homeassistant/components/refoss/coordinator.py

```diff
@@ -0,0 +1,277 @@
+
+from __future__ import annotations
+
+import asyncio
+from asyncio import AbstractEventLoop
+from collections.abc import Iterable
+from datetime import timedelta
+from functools import lru_cache
+from typing import Optional, TypeVar
+
+import async_timeout
+from refoss_ha.const import DOMAIN, LOGGER, PUSH
+from refoss_ha.controller.device import BaseDevice
+from refoss_ha.controller.system import SystemAllMixin
+from refoss_ha.controller.toggle import ToggleXMix
+from refoss_ha.enums import Namespace
+from refoss_ha.http_device import HttpDeviceInfo
+from refoss_ha.socket_util import SocketUtil, pushStateDataList
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+T = TypeVar("T", bound=BaseDevice)
+
+_ABILITY_MATRIX = {
+    Namespace.CONTROL_TOGGLEX.value: ToggleXMix,
+    Namespace.SYSTEM_ALL.value: SystemAllMixin,
+}
+
+
+class RefossCoordinator(DataUpdateCoordinator):
+    """RefossCoordinator ."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        update_interval: timedelta,
+        loop: Optional[AbstractEventLoop] = None,
+    ) -> None:
+        """__init__ ."""
+        self._devices_by_internal_id: dict[str, BaseDevice] = {}
+        self._setup_done = False
+        self.socket = SocketUtil()
+        self.tasks: list[asyncio.Task] = []
+        self._loop = asyncio.get_event_loop() if loop is None else loop
+
+        super().__init__(
+            hass=hass,
+            logger=LOGGER,
+            name="refoss_coordinator",
+            update_interval=update_interval,
+            update_method=self._async_fetch_data,
+        )
+
+    async def initial_setup(self):
+        """initial_setup ."""
+        if self._setup_done:
+            raise ValueError("This coordinator was already set up")
+
+        # Listening for socket messages
+        self.socket.startReveiveMsg()
+
+        task = asyncio.create_task(self.HandlePushState())
+        self.tasks.append(task)
+        task.done()
+
+        devicelist = self.socket.async_socket_find_devices()
+        self.async_set_updated_data({device.uuid: device for device in devicelist})
+
+        await self.async_device_discovery(cached_http_device_list=devicelist)
+
+        self._setup_done = True
+        LOGGER.info("Initial_setup ok")
+
+    async def _async_fetch_data(self):
+        """_async_fetch_data ."""
+        async with async_timeout.timeout(10):
+            devices = self.socket.async_socket_find_devices()
+            return {device.uuid: device for device in devices}
+
+    def find_devices(self, device_uuids: Optional[Iterable[str]] = None) -> list[T]:
+        """find_devices ."""
+        res = self._devices_by_internal_id.values()
+        list_res = list(res)
+        if device_uuids is not None:
+            list_res = [d for d in list_res if d.uuid in device_uuids]
+
+        return list_res
+
```

Comment:
For local devices, we want one config entry per device so we can retry by raising `ConfigEntryNotReady` and handle ip changes via discovery flows